Windows fails to link sandbox, Visual Studio 2022

This is a placeholder issue. It would be nice to nice easy instructions to install Windows tools.

Previous working setup is described in the Scala Native docs using Visual Studio Community 2019 and LLVM vendored version 12.0.1.

Current library setup is Visual Studio Build Tools 2022, version 17.4.3

Scala Native HEAD checked out.

Details:
Edition	Windows 10 Enterprise
Version	22H2
Installed on	‎5/‎26/‎2022
OS build	19045.2364
Experience	Windows Feature Experience Pack 120.2212.4190.0

Windows installed Clang/LLVM
```shell
> clang --version
clang version 15.0.1
Target: i686-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\Llvm\bin
```
```scala
[info] Checking intermediate code (475 ms)
[info] Dumping intermediate code (linked) (1301 ms)
[info] Discovered 627 classes and 4191 methods
[info] Optimizing (debug mode) (2267 ms)
[info] Checking intermediate code (109 ms)
[info] Dumping intermediate code (optimized) (550 ms)
[info] Dumping intermediate code (lowered) (834 ms)
[info] Generating intermediate code (3338 ms)
[info] Produced 36 files
[info] Compiling to native code (59069 ms)
[info] channels.ll.o : error LNK2019: unresolved external symbol _WriteFile referenced in function __SM33java.nio.channels.FileChannelImplD5writeLAb_iiuEO
[info]   Hint on symbols that are defined and could potentially match:
[info]     _WriteFile@20
[info] file.ll.o : error LNK2019: unresolved external symbol _GetLastError referenced in function __SM31java.nio.file.WindowsException$D6onPathL16java.lang.StringL19java.io.IOExceptionEO
[info]   Hint on symbols that are defined and could potentially match:
[info]     _GetLastError@0
[info] windows.ll.o : error LNK2019: unresolved external symbol _GetStdHandle referenced in function __SM40scala.scalanative.windows.ConsoleApiExt$D5stdInL28scala.scalanative.unsafe.PtrEO
[info]   Hint on symbols that are defined and could potentially match:
[info]     _GetStdHandle@4
[info] windows.ll.o : error LNK2019: unresolved external symbol _FormatMessageW referenced in function __SM46scala.scalanative.windows.ErrorHandlingApiOps$D23$anonfun$errorMessage$1L31scala.scalanative.unsigned.UIntL29scala.scalanative.unsafe.ZoneL16java.lang.StringEPT46scala.scalanative.windows.ErrorHandlingApiOps$
[info] C:\Users\ekr32456\workspace\scala-native\sandbox\.2.12\target\scala-2.12\sandbox.exe : fatal error LNK1120: 4 unresolved externals
[error] clang++: error: linker command failed with exit code 1120 (use -v to see invocation)
[error] Failed to link C:\Users\<me>\workspace\scala-native\sandbox\.2.12\target\scala-2.12\sandbox.exe
[error] (sandbox2_12 / Compile / nativeLink) Failed to link C:\Users\<me>\workspace\scala-native\sandbox\.2.12\target\scala-2.12\sandbox.exe
[error] Total time: 327 s (05:27), completed Jan 6, 2023, 8:53:01 PM
```

I just updated to Visual Studio 2022 17.7.4 which has clang 16.0.5. Still no luck.

I found this which seems to at least explain the problem of external symbols.
https://www.plantation-productions.com/Webster/Win32Asm/win32API.html

```
[info] Compiling to native code (47307 ms)
[info] windows.ll.o : error LNK2019: unresolved external symbol _GetStdHandle referenced in function __SM40scala.scalanative.windows.ConsoleApiExt$D5stdInL28scala.scalanative.unsafe.PtrEO
[info]   Hint on symbols that are defined and could potentially match:
[info]     _GetStdHandle@4
[info] windows.ll.o : error LNK2019: unresolved external symbol _FormatMessageW referenced in function __SM46scala.scalanative.windows.ErrorHandlingApiOps$D23$anonfun$errorMessage$1L31scala.scalanative.unsigned.UIntL29scala.scalanative.unsafe.ZoneL16java.lang.StringEPT46scala.scalanative.windows.ErrorHandlingApiOps$
[info] util.ll.o : error LNK2019: unresolved external symbol _CloseHandle referenced in function __SM31java.util.WindowsHelperMethods$D13withUserTokenL31scala.scalanative.unsigned.UIntL15scala.Function1L16java.lang.ObjectEO
[info]   Hint on symbols that are defined and could potentially match:
[info]     _CloseHandle@4
[info] util.ll.o : error LNK2019: unresolved external symbol _GetCurrentThread referenced in function "?catch$36@?0?_SM31java.util.WindowsHelperMethods$D13withUserTokenL31scala.scalanative.unsigned.UIntL15scala.Function1L16java.lang.ObjectEO@4HA" (?catch$36@?0?_SM31java.util.WindowsHelperMethods$D13withUserTokenL31scala.scalanative.unsigned.UIntL15scala.Function1L16java.lang.ObjectEO@4HA)
[info]   Hint on symbols that are defined and could potentially match:
[info]     _GetCurrentThread@0
[info] util.ll.o : error LNK2019: unresolved external symbol _OpenThreadToken referenced in function "?catch$36@?0?_SM31java.util.WindowsHelperMethods$D13withUserTokenL31scala.scalanative.unsigned.UIntL15scala.Function1L16java.lang.ObjectEO@4HA" (?catch$36@?0?_SM31java.util.WindowsHelperMethods$D13withUserTokenL31scala.scalanative.unsigned.UIntL15scala.Function1L16java.lang.ObjectEO@4HA)
[info] util.ll.o : error LNK2019: unresolved external symbol _GetCurrentProcess referenced in function "?catch$36@?0?_SM31java.util.WindowsHelperMethods$D13withUserTokenL31scala.scalanative.unsigned.UIntL15scala.Function1L16java.lang.ObjectEO@4HA" (?catch$36@?0?_SM31java.util.WindowsHelperMethods$D13withUserTokenL31scala.scalanative.unsigned.UIntL15scala.Function1L16java.lang.ObjectEO@4HA)
[info]   Hint on symbols that are defined and could potentially match:
[info]     _GetCurrentProcess@0
[info] util.ll.o : error LNK2019: unresolved external symbol _OpenProcessToken referenced in function "?catch$36@?0?_SM31java.util.WindowsHelperMethods$D13withUserTokenL31scala.scalanative.unsigned.UIntL15scala.Function1L16java.lang.ObjectEO@4HA" (?catch$36@?0?_SM31java.util.WindowsHelperMethods$D13withUserTokenL31scala.scalanative.unsigned.UIntL15scala.Function1L16java.lang.ObjectEO@4HA)
[info] file.ll.o : error LNK2019: unresolved external symbol _GetLastError referenced in function __SM31java.nio.file.WindowsException$D6onPathL16java.lang.StringL19java.io.IOExceptionEO
[info]   Hint on symbols that are defined and could potentially match:
[info]     _GetLastError@0
[info] channels.ll.o : error LNK2019: unresolved external symbol _WriteFile referenced in function __SM33java.nio.channels.FileChannelImplD10writeArrayLAb_iiiEPT33java.nio.channels.FileChannelImpl
[info]   Hint on symbols that are defined and could potentially match:
[info]     _WriteFile@20
[info] channels.ll.o : error LNK2019: unresolved external symbol _SetFilePointerEx referenced in function __SM33java.nio.channels.FileChannelImplD7seekEOFuEPT33java.nio.channels.FileChannelImpl
[info]   Hint on symbols that are defined and could potentially match:
[info]     _SetFilePointerEx@20
[info] lang.ll.o : error LNK2019: unresolved external symbol _GetTempPathA referenced in function __SM17java.lang.System$D14loadPropertiesL20java.util.PropertiesEPT17java.lang.System$
[info] lang.ll.o : error LNK2019: unresolved external symbol _GetLocaleInfoEx referenced in function __SM17java.lang.System$D17currentTimeMillisjEO
[info] lang.ll.o : error LNK2019: unresolved external symbol _GetCurrentDirectoryW referenced in function __SM17java.lang.System$D17systemProperties0L20java.util.PropertiesEPT17java.lang.System$
[info] lang.ll.o : error LNK2019: unresolved external symbol _GetUserProfileDirectoryW referenced in function __SM17java.lang.System$D31$anonfun$getUserHomeDirectory$1L28scala.scalanative.unsafe.PtrL12scala.OptionEPT17java.lang.System$
[info] lang.ll.o : error LNK2019: unresolved external symbol _CreateEventW referenced in function __SM17java.lang.Thread$RE
[info] C:\Users\<me>\workspace\scala-native\sandbox\.2.13\target\scala-2.13\native\Test.exe : fatal error LNK1120: 15 unresolved externals
[error] clang++: error: linker command failed with exit code 1120 (use -v to see invocation)
[info] Total (66282 ms)
[error] Failed to link C:\Users\<me>\workspace\scala-native\sandbox\.2.13\target\scala-2.13\native\Test.exe
[error] (sandbox2_13 / Compile / nativeLink) Failed to link C:\Users\<me>\workspace\scala-native\sandbox\.2.13\target\scala-2.13\native\Test.exe
[error] Total time: 81 s (01:21), completed Sep 15, 2023, 12:31:01 PM
sbt:scala-native>
```

Update: I have been able to compile 64 bit using the LLVM toolchain directly.
Install into `C:\Users\<me>\AppData\Local\LLVM`
```
$ ~/AppData/Local/LLVM/bin/clang --version
clang version 16.0.6
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Users\<me>\AppData\Local\LLVM\bin
```
Also using the Visual Studio toolchain 64 bit using target triple `x86_64-pc-windows-msvc` with version appended to avoid warnings `19.37.32824`
```
$ clang --version
clang version 16.0.5
Target: i686-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\Llvm\bin
```
As for 32 bit with a bit of Aly 's help I get close. Config in `sandbox`. Instead of the link errors shown above I added `-fuse-ld=lld-link` and then I got
```
[info] Compiling to native code (1223 ms)
[error] lld-link: error: duplicate symbol: __Sinh
[error] >>> defined at minkernel\crts\ucrt\src\appcrt\tran\_csinh.h:115
[error] >>>            libucrt.lib(csinh.obj)
[error] >>> defined at libcpmt.lib(xsinh.obj)
[error] lld-link: error: duplicate symbol: __FSinh
[error] >>> defined at minkernel\crts\ucrt\src\appcrt\tran\_csinh.h:115
[error] >>>            libucrt.lib(csinhf.obj)
[error] >>> defined at libcpmt.lib(xfsinh.obj)
[error] clang++: error: linker command failed with exit code 1 (use -v to see invocation)
[info] Total (21246 ms)
```
Then I found that @WojciechMazur had already been down this path and put code link this in the link command so then I added the additional link option.
```scala
.settings {
        nativeConfig ~= (
          _.withTargetTriple("i686-pc-windows-msvc19.37.32824")//19.37.32824
          .withCompileOptions(_ ++ Seq("-m32"))
          .withLinkingOptions(_ ++ Seq("-m32", "-fuse-ld=lld-link", "-Wl,/force:multiple"))
        )
      }
```
But with this option added I get this while linking which is in the compiler code.
```
[info] Total (6475 ms)
[info] Generating intermediate code (with debug metadata) (1779 ms)
[error] java.util.concurrent.ExecutionException: Boxed Error
[error]         at scala.concurrent.impl.Promise$.resolver(Promise.scala:87)
[error]         at scala.concurrent.impl.Promise$.scala$concurrent$impl$Promise$$resolveTry(Promise.scala:79)
[error]         at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:284)
[error]         at scala.concurrent.Promise.complete(Promise.scala:53)
[error]         at scala.concurrent.Promise.complete$(Promise.scala:52)
[error]         at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:187)
[error]         at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
[error]         at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
[error]         at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
[error]         at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
[error]         at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
[error]         at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
[error]         at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
[error]         at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
[error] Caused by: java.lang.AssertionError: assertion failed
[error]         at scala.Predef$.assert(Predef.scala:208)
[error]         at scala.scalanative.codegen.llvm.CodeGen$.$anonfun$emit$10(CodeGen.scala:112)
[error]         at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:659)
[error]         at scala.util.Success.$anonfun$map$1(Try.scala:255)
[error]         at scala.util.Success.map(Try.scala:213)
[error]         at scala.concurrent.Future.$anonfun$map$1(Future.scala:292)
[error]         at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)
[error]         at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)
[error]         at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)
[error]         at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
[error]         at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
[error]         at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
[error]         at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
[error]         at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
[error]         at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
[error] (sandbox2_13 / Compile / nativeLink) java.util.concurrent.ExecutionException: Boxed Error
```

This actually looks like some progress for `0.5.0-RC3`.
```sh
$ clang --version
clang version 16.0.5
Target: i686-pc-windows-msvc
Thread model: posix
InstalledDir: C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\Llvm\bin
```
```sh
[info] Linking (multithreadingEnabled=true, disable if not used) (1422 ms)
[info] Dumping intermediate code (linked) (822 ms)
[info] Discovered 853 classes and 5172 methods after classloading
[info] Checking intermediate code (full) (96 ms)
[info] Multithreading was not explicitly enabled - initial class loading has not detected any usage of system threads. Multithreading support will be disabled to improve performance.
[info] Linking (multithreadingEnabled=false) (1253 ms)
[info] Dumping intermediate code (linked) (209 ms)
[info] Discovered 514 classes and 2562 methods after classloading
[info] Checking intermediate code (full) (22 ms)
[info] Dumping intermediate code (optimized) (436 ms)
[info] Discovered 497 classes and 1949 methods after optimization
[info] Checking intermediate code (full) (26 ms)
[info] Optimizing (debug mode) (1995 ms)
[info] Dumping intermediate code (lowered) (1043 ms)
[info] Produced 44 LLVM IR files
[info] Generating intermediate code (with debug metadata) (41515 ms)
[error] C:/Users/<user>/workspace/scala-native/sandbox/.3/target/scala-3.4.1/native/dependencies/nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/gc/immix/ImmixGC.c:144:25: error: incompatible function pointer types passing 'ThreadStartRoutine' (aka 'unsigned long (*)(void *)') to parameter of type 'LPTHREAD_START_ROUTINE' (aka 'unsigned long (*)(void *) __attribute__((stdcall))') [-Wincompatible-function-pointer-types]
[error]                         (ThreadStartRoutine)&ProxyThreadStartRoutine,
[error]                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[error] C:\Program Files (x86)\Windows Kits\10\Include\10.0.22000.0\um\processthreadsapi.h:184:33: note: passing argument to parameter 'lpStartAddress' here
[error]     _In_ LPTHREAD_START_ROUTINE lpStartAddress,
[error]                                 ^
[error] 1 error generated.
[info] Compiling to native code (125786 ms)
[info] Total (162366 ms)
[error] stack trace is suppressed; run 'last sandbox3 / Compile / nativeLink' for the full output
[error] stack trace is suppressed; run 'last sandbox3 / Compile / nativeLink' for the full output
[error] stack trace is suppressed; run 'last sandbox3 / Compile / nativeLink' for the full output
[error] stack trace is suppressed; run 'last sandbox3 / Compile / nativeLink' for the full output
[error] stack trace is suppressed; run 'last sandbox3 / Compile / nativeLink' for the full output
[error] stack trace is suppressed; run 'last sandbox3 / Compile / nativeLink' for the full output
[error] stack trace is suppressed; run 'last sandbox3 / Compile / nativeLink' for the full output
[error] stack trace is suppressed; run 'last sandbox3 / Compile / nativeLink' for the full output
[error] Failed to compile C:/Users/<user>/workspace/scala-native/sandbox/.3/target/scala-3.4.1/native/dependencies/nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/gc/immix/ImmixGC.c
[error] (sandbox3 / Compile / nativeLink) Failed to compile C:/Users/<user>/workspace/scala-native/sandbox/.3/target/scala-3.4.1/native/dependencies/nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/gc/immix/ImmixGC.c
```

--- a/nativelib/src/main/resources/scala-native/gc/shared/ScalaNativeGC.h
+++ b/nativelib/src/main/resources/scala-native/gc/shared/ScalaNativeGC.h
@@ -14,10 +14,12 @@
 typedef DWORD ThreadRoutineReturnType;
 #else
 #include <pthread.h>
+// define as nothing on non Windows
+#define WINAPI
 typedef void *ThreadRoutineReturnType;
 #endif
-
-typedef ThreadRoutineReturnType (*ThreadStartRoutine)(void *);
+// Requires WINAPI which is defined as __stdcall on Windows
+typedef ThreadRoutineReturnType(WINAPI *ThreadStartRoutine)(void *);
 typedef void *RoutineArgs;
 
 void scalanative_GC_init();

